src/java.base/share/classes/java/util/HashMap.java

```diff
@@ -500,7 +500,7 @@ public HashMap(Map<? extends K, ? extends V> m) {
      * true (relayed to method afterNodeInsertion).
      */
     final void putMapEntries(Map<? extends K, ? extends V> m, boolean evict) {
-        int s = m.size();
+        int s = Math.max(size() + m.size(), m.size());
```

Comment:
If we decide this approach is best, shouldn't we do the same for ConcurrentHashMap?

Reply:
Sure. Created https://bugs.openjdk.org/browse/JDK-8324798. Won't implement that until this gets merged.

Reply:
Circling back to "the case where many keys exist in both maps".  What are your thoughts on some optimized variant/refactoring of:
`int s = Math.max(m.size() - this.size(), 1);`

This would make it so we aggressively presize when know a large map is being added to a small or empty map.  Obviously, if both maps are large and don't contain the same keys this doesn't help.

Also, are we sure it is wise to actually call m.size() when we don't know the runtime of the size operation on the given map?  The only thing we can reason about is runtime of this.size().

Reply:
> Circling back to "the case where many keys exist in both maps". What are your thoughts on some optimized variant/refactoring of:
int s = Math.max(m.size() - this.size(), 1);

Calling `Math.max` is unnecessary here. Its ok if `s` ends up non-positive, presizing would become a no-op in that case.

I don't understand why there is a subtraction. If `m` is a larger map than `this`, and has all of `this`'s keys, the resulting map will have the same size as `m`. In this case, we should just presize to `m.size()`.

> Also, are we sure it is wise to actually call m.size() when we don't know the runtime of the size operation on the given map? 

Not sure, but the current code already calls `m.size()`. I can create a `int msize = m.size()` so we only call it once.